formula1_race_analysis/q1_session_analyzer.py

```diff
@@ -0,0 +1,161 @@
+        if driver.start_time is None or driver.end_time is None:
+            raise CustomMissedTimestampError(
+                f"Error! Missing timestamp for driver: '{driver.driver_name}'. Start or end time is not provided.",
+            )
+        start_time_in_seconds = convert_timestamp_to_seconds(driver.start_time)
+        end_time_in_seconds = convert_timestamp_to_seconds(driver.end_time)
+
+        race_duration_seconds = end_time_in_seconds - start_time_in_seconds
+        if start_time_in_seconds > end_time_in_seconds:
+            raise CustomInvalidRaceTimeError(
+                f"Race time error for driver: '{driver.driver_name}'. Start time is greater than end time.",
+            )
+        driver.race_duration = race_duration_seconds
+
+
+def convert_timestamp_to_seconds(timestamp: str) -> float:
+    """
+    Converts the timestamp format "YYYY-MM-DD_HH:MM:SS.sss" to a total number of seconds.
+    If the timestamp isn't in the correct format, a CustomInvalidFormatDataError raises.
+    """
+    try:
+        dt = datetime.strptime(timestamp, "%Y-%m-%d_%H:%M:%S.%f")
+        hours = dt.hour
+        minutes = dt.minute
+        seconds = dt.second
+        milliseconds = dt.microsecond / 1000000
+        total_seconds = hours * SECONDS_PER_HOUR + minutes * SECONDS_PER_MINUTE + seconds + milliseconds
+    except ValueError as error:
+        raise CustomInvalidFormatDataError(
+            f"The timestamp format: '{timestamp}' is incorrect. Expected format: YYYY-MM-DD_HH:MM:SS.sss",
+        ) from error
+    return total_seconds
+
+
+def get_q1_driver_position(database: list) -> list:
+    """
+    This function sorts the input list of drivers by their race duration in ascending order.
+    Then assigns qualifying positions (Q1) to drivers based on their race durations.
+    """
+    sorted_driver_database = sorted(database, key=lambda driver: driver.race_duration)
+    for position, driver in enumerate(sorted_driver_database, start=1):
+        driver.q1_position = position
+    return sorted_driver_database
+
+
+def display_race_report(database: list[Driver]) -> list:
+    """
+    Generates a formatted race report
+    """
+    report = []
+    for driver in database:
+        formatted_race_duration = format_race_duration(driver.race_duration)
+        report.append(f"{driver.q1_position}. {driver.driver_name} | {driver.car_model} | {formatted_race_duration}")
+        if driver.q1_position == THE_NUMBER_OF_FASTEST_DRIVERS_PASSED_Q1:
+            report.append("___________________________________________________________")
+            continue
+    return report
+
+
+def format_race_duration(race_duration_in_seconds: float) -> str:
+    """
+    Formats the race duration in seconds into MM:SS.mmm format.
+    """
+    race_minutes = int(race_duration_in_seconds // SECONDS_PER_MINUTE)
+    race_seconds = race_duration_in_seconds - race_minutes * SECONDS_PER_MINUTE
+    return f"{race_minutes}:{race_seconds:06.3f}"
```

Comment:
I would suggest bringing some more folder structure into your package. Get some core/ , display/ folders put formatting  functions in the respective file in folder that makes sense